File: Blind 75/13 - 1D Dynamic Programming/213-house-robber-II.py
'''
You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed. All houses at this place are arranged in a circle. That means the first house is the neighbor of the last one. Meanwhile, adjacent houses have a security system connected, and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
'''
from time import time
from typing import List


class Solution:
    def rob_helper(self, nums: List[int]) -> int:
        prev, curr = 0, 0
        for n in nums:
            prev, curr = curr, max(prev + n, curr)
        return curr

    def rob(self, nums: List[int]) -> int:
        return max(
            nums[0], self.rob_helper(nums[1:]), self.rob_helper(nums[:-1])
        )

    def reference_helper(self, nums: List[int]) -> int:
        plan_a, plan_b = 0, 0

        for n in nums:
            temp = max(n + plan_a, plan_b)
            plan_a = plan_b
            plan_b = temp

        return plan_b

    def reference(self, nums: List[int]) -> int:
        return max(
            nums[0],
            self.reference_helper(nums[1:]),
            self.reference_helper(nums[:-1]),
        )

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.rob(case))
                else:
                    self.rob(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [2, 3, 2],
        [1, 2, 3, 1],
        [1, 2, 3],
        # Additional
        [1],
    ]
    test.quantify(test_cases)

File: Blind 75/13 - 1D Dynamic Programming/JavaScript/213-house-robber-II.js
const { performance } = require("perf_hooks");

class Solution {
  method(nums) {
    const robHelper = (nums) => {
      let [prev, curr] = [0, 0];
      for (const n of nums) {
        [prev, curr] = [curr, Math.max(curr, prev + n)];
      }
      return curr;
    };

    return Math.max(
      nums[0],
      robHelper(nums.slice(1)),
      robHelper(nums.slice(0, -1))
    );
  }

  reference(nums) {
    const isBaseCase1 = nums.length === 0;
    if (isBaseCase1) return 0;

    const isBaseCase2 = nums.length === 1;
    if (isBaseCase2) return nums[0];

    const left = this.search(nums, 0, nums.length - 2); /* Time O(N) */
    const right = this.search(nums, 1, nums.length - 1); /* Time O(N) */

    return Math.max(left, right);
  }

  search = (nums, start, end) => {
    let [left, mid] = [0, 0];

    for (let i = start; i <= end; i++) {
      /* Time O(N) */
      const temp = mid;
      const right = nums[i];
      const house = left + right;

      mid = Math.max(mid, house);
      left = temp;
    }

    return mid;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [2, 3, 2],
  [1, 2, 3, 1],
  [1, 2, 3],
];
test.quantify(testCases);
